### PR TITLE
Convert LinearAlgebra::distributed::Vector to Kokkos 

### PR DESCRIPTION
**Pull request by masterleinad**
Depends on #14510.

### CODE REVIEW

**Comment by tjhei** [comment]
Can you rebase to see if the new MPI Jenkins passes?

File: include/deal.II/lac/vector_operations_internal.h [hunk 1]
```diff
@@ -2472,32 +2478,22 @@ namespace internal
...
+        auto exec = typename ::dealii::MemorySpace::Default::kokkos_space::
+          execution_space{};
+        Kokkos::parallel_reduce(
+          Kokkos::RangePolicy<
+            ::dealii::MemorySpace::Default::kokkos_space::execution_space>(
+            exec, 0, size),
+          KOKKOS_LAMBDA(size_type i, Number & update) {
+            update += data.values(i);
+          },
+          result);
```
Comment (Rombur) [inline include/deal.II/lac/vector_operations_internal.h:2494]: The name of the function is `mean_value`, shouldn't we divide the `result` by the size?

File: include/deal.II/lac/vector_operations_internal.h [hunk 2]
```diff
@@ -2508,44 +2504,55 @@ namespace internal
...
+          Kokkos::RangePolicy<
+            ::dealii::MemorySpace::Default::kokkos_space::execution_space>(
+            exec, 0, size),
+          KOKKOS_LAMBDA(size_type i, Number & update) {
+#if KOKKOS_VERSION < 30400
+            update += fabs(data.values(i));
+#elif KOKKOS_VERSION < 30700
+            update += Kokkos::Experimental::fabs(data.values(i));
+#else
+            update += Kokkos::fabs(data.values(i));
```
Comment (Rombur) [inline include/deal.II/lac/vector_operations_internal.h:2522]:
```suggestion
            update += Kokkos::abs(data.values(i));
```

File: include/deal.II/lac/vector_operations_internal.h [hunk 3]
```diff
@@ -2508,44 +2504,55 @@ namespace internal
...
+            ::dealii::MemorySpace::Default::kokkos_space::execution_space>(
+            exec, 0, size),
+          KOKKOS_LAMBDA(size_type i, Number & update) {
+#if KOKKOS_VERSION < 30400
+            update += pow(fabs(data.values(i)), exp);
+#elif KOKKOS_VERSION < 30700
+            update += Kokkos::Experimental::pow(
+              Kokkos::Experimental::fabs(data.values(i)), exp);
+#else
+            update += Kokkos::pow(Kokkos::fabs(data.values(i)), exp);
```
Comment (Rombur) [inline include/deal.II/lac/vector_operations_internal.h:2552]:
```suggestion
            update += Kokkos::pow(Kokkos::abs(data.values(i)), exp);
```

File: include/deal.II/lac/vector_operations_internal.h [hunk 4]
```diff
@@ -2625,27 +2627,26 @@ namespace internal
...
-            cudaError_t cuda_error_code = cudaMemcpy(data.values.data(),
-                                                     v_data.values.data(),
-                                                     size * sizeof(Number),
-                                                     cudaMemcpyHostToDevice);
-            AssertCuda(cuda_error_code);
+            Kokkos::deep_copy(
+              Kokkos::subview(data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)),
+              Kokkos::subview(v_data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)));
```
Comment (Rombur) [inline include/deal.II/lac/vector_operations_internal.h:2642]: Don't think you need the subview

File: source/lac/la_parallel_vector_cuda.cc [hunk 5]
```diff
@@ -24,40 +24,50 @@ namespace LinearAlgebra
 {
   namespace distributed
   {
-    template class Vector<float, ::dealii::MemorySpace::CUDA>;
-    template class Vector<double, ::dealii::MemorySpace::CUDA>;
+    template class Vector<float, ::dealii::MemorySpace::Default>;
+    template class Vector<double, ::dealii::MemorySpace::Default>;
```
Comment (Rombur) [inline source/lac/la_parallel_vector_cuda.cc:28]: Is this file still necessary? We could do this in `la_parallel_vector.inst.in`

File: include/deal.II/lac/vector_operations_internal.h [hunk 6]
```diff
@@ -2111,103 +2110,98 @@ namespace internal
...
-            cudaError_t cuda_error_code = cudaMemcpy(data.values.data(),
-                                                     v_data.values.data(),
-                                                     size * sizeof(Number),
-                                                     cudaMemcpyDeviceToHost);
-            AssertCuda(cuda_error_code);
+            Kokkos::deep_copy(
+              Kokkos::subview(data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)),
+              Kokkos::subview(v_data.values,
+                              Kokkos::pair<size_type, size_type>(0, size)));
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/lac/vector_operations_internal.h:2122]: I don't think this would be correct. We use this function to copy only the locally owned part of the vector.

**Comment by Rombur** [comment]
Code doesn't compile anymore. Did you mean to push the new commits from Lauren's account?

**Comment by masterleinad** [comment]
> Code doesn't compile anymore. Did you mean to push the new commits from Lauren's account?

I fixed the authorship for all commits and reverted unifying the implementations for `import_elements`. We need the specialization as long as we don't replace `TBB` with `Kokkos` which I would prefer to fo somewhere else. All your other comments should be addressed.

**Comment by masterleinad** [comment]
It seems `Kokkos::Experimental::abs` didn't exist (but `Kokoks::abs` does now).

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 7]
```diff
@@ -372,108 +375,124 @@ namespace LinearAlgebra
...
+                                                                   n_elements),
+            KOKKOS_LAMBDA(size_type i) {
+              indices[i] = locally_owned_elem.index_within_set(
+                tmp_index_set.nth_index_in_set(i));
+            });
+          host_exec.fence();
+          Kokkos::realloc(indices_dev, tmp_n_elements);
+          Kokkos::deep_copy(indices_dev,
+                            Kokkos::View<size_type *>(indices.data(),
+                                                      tmp_n_elements));
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h:443]: `indices` is already a `View` so why do you create a new one? Shouldn't this be a subview?

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 8]
```diff
@@ -372,108 +375,124 @@ namespace LinearAlgebra
...
-                                  result_device,
-                                  sizeof(Number),
-                                  cudaMemcpyDeviceToHost);
-          AssertCuda(error_code);
-          // Free the memory on the device
-          error_code = cudaFree(result_device);
-          AssertCuda(error_code);
+          typename ::dealii::MemorySpace::Default::kokkos_space::execution_space
+            exec;
+          Kokkos::parallel_reduce(
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h:479]: You forgot to use a name for all the `parallel_reduce`

**Review by Rombur — COMMENTED** [review]
Just minor comments. Do you know if this PR fixes the problem we have with ghosted vector (see the cuda/parallel_vector tests failing [here](https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=958))?

**Comment by masterleinad** [comment]
> Just minor comments. Do you know if this PR fixes the problem we have with ghosted vector (see the cuda/parallel_vector tests failing [here](https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=958))?

I haven't run the tests on NVIDIA GPUs but only on Intel GPUs (and CPUs using `SYCL`) and those tests were/are passing there.

**Comment by Rombur** [comment]
Maybe a problem when we use the CUDA-aware MPI path then. Is there something similar to CUDA-aware MPI with SYCL?

**Comment by masterleinad** [comment]
Yes, there is GPU-aware MPI for Intel GPUs but this pull request doesn't address that. That's what #14571 is for (I have only tested that one with `Serial` and `SYCL`+CPU).

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 7 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h]: Fixed.

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 8 above)
Comment (masterleinad) [reply include/deal.II/lac/la_parallel_vector.templates.h:480]: Fixed.

**Review by Rombur — APPROVED** [review]
